draft-jmiller-json-web-proof.md

```diff
@@ -121,44 +98,27 @@ For example:
+
 ## Payloads
 
-Payloads are always represented as an ordered array.  Each payload's content type is either known by context or specified in that payload's associated header.  The default content type is `application/json`, with the JSON object values being JWT claims.
+Payloads are always represented as an ordered array.  The mapping of which value is in which payload slot is out of scope of this specification.
 
-Example payloads:
-```json
-{
-    "given_name":"Jane",
-    "family_name":"Doe"
-}
-```
-```json
-{
-    "email":"janedoe@example.com"
-}
-```
-```json
-{
-    "birthdate":"0000-03-22"
-}
-```
+In order to support ZKPs, individual payloads cannot be serialized before they are passed into an algorithm implementation.  This enables the algorithms to accept and internally encode elliptic curve points, blinded values, plain numbers, membership keys, etc.  Implementations are therefore required to provide optional arguments for each payload such that the application can utilize these capabilities as needed.
 
+When selective disclosure preferences are applied, any one or more payloads may be hidden.  The position of other payloads does not change due to any preceeding ones being hidden - the resulting array will simply be sparse without the hidden payloads.
 
 ## Proof
 
 The proof value is a binary octet string that is opaque to applications.  Individual proof-supporting algorithms are responsible for the contents and security of the proof value along with any required internal structures to it.
 
-All proofs MUST include integrity protection of the JWP's base64url encoded header value.  This value cannot be hidden and the protection MUST be included in the proof even after derivation.
-
-# Derivation
-
-When selective disclosure preferences are applied to a derivation, any one or more payloads may be hidden.  The position of other payloads does not change due to any proceeding ones being hidden; the resulting array will simply be sparse, with the hidden payloads being replaced by the value `null`.
+Implementations will also be required to provide optional arguments for each payload as input into the `prove` step.  These arguments can be used for generating predicate proofs, linking options, etc.
 
-Algorithms SHOULD generate a new un-correlatable proof value when a JWP is derived.  A JWP may also be single-use, where correlation across multiple derivations is not a factor.
+Algorithms SHOULD generate a new un-correlatable proof value during the `prove` step.  A JWP may also be single-use, where correlation across multiple derivations is not a factor.
 
 # Serializations
 
-Each payload MUST be base64url encoded when preparing it to be serialized.  The header and proof are also individually base64url encoded.
+Each disclosed payload MUST be base64url encoded when preparing it to be serialized.  The header and proof are also individually base64url encoded.
```

Comment:
As much as I love base64url... it kinda sucks that `-` breaks highlighting... I take it this is one of those cases where the old way of doing things is probably worth preserving more than the improvement, buts its worth noting that base64url inflates data, has no built in error correction, is not "double clickable" and allows multiple serializations to deserialize to the same buffer.

These issues might not be worth deviating from the convention of just using bas64url, but they are worth noting.

Reply:
I disagree with the "multiple serializations to deserialize to the same buffer" comment.  There is a single base64url encoding of any string of bytes.  Given there's a 1:1 correspondence, I don't understand this comment, @OR13 .

Reply:
https://github.com/decentralized-identity/sidetree/blob/3184bfabd4b01e22c9d56fb16feadb5b303da9c4/lib/core/versions/1.0/Multihash.ts#L208

Reply:
Given that we plan on defining a COSE Proof analog to JWP, my preference is to keep JWP as close to JOSE conventions as possible for easier adoption in those existing environments.

COSE Proofs can serve as the basis for much more efficient representations that can be serialized/encoded independently by a constrained transport as needed.